File: Iniciante/1040 - Média 3/Python/1040.py
import math

WEIGHTS = [2, 3, 4, 1]


def main():
    grades = list(map(float, input().split()))
    average = calculate_weighted_average(grades)
    average = math.floor(average * 10) / 10.0
    print(f"Media: {average:.1f}")
    check_student_status(average)


def calculate_weighted_average(grades):
    total = 0.0
    total_weight = 0.0

    for grade, weight in zip(grades, WEIGHTS):
        total += grade * weight
        total_weight += weight

    return total / total_weight


def handle_exam(initial_average):
    exam_grade = float(input())
    print(f"Nota do exame: {exam_grade:.1f}")

    final_average = (initial_average + exam_grade) / 2.0

    if final_average >= 5.0:
        print("Aluno aprovado.")
    else:
        print("Aluno reprovado.")

    print(f"Media final: {final_average:.1f}")


def check_student_status(average):
    if average >= 7.0:
        print("Aluno aprovado.")
    elif average < 5.0:
        print("Aluno reprovado.")
    else:
        print("Aluno em exame.")
        handle_exam(average)


if __name__ == "__main__":
    main()

File: Iniciante/1040 - Média 3/TypeScript/1040.ts
import * as fs from "fs";

const WEIGHTS: number[] = [2, 3, 4, 1];

function main(): void {
  const input: string = fs.readFileSync("/dev/stdin", "utf8");
  const lines: string[] = input.trim().split("\n");
  const grades: number[] = lines[0].trim().split(" ").map(Number);
  let average: number = calculateWeightedAverage(grades);
  average = truncateDecimal(average);
  console.log(`Media: ${average.toFixed(1)}`);
  if (average >= 7.0) {
    console.log("Aluno aprovado.");
  } else if (average < 5.0) {
    console.log("Aluno reprovado.");
  } else {
    console.log("Aluno em exame.");
    const examGrade: number = parseFloat(lines[1]);
    processExam(average, examGrade);
  }
}

function calculateWeightedAverage(grades: number[]): number {
  const sumOfWeights = WEIGHTS.reduce((acc, weight) => acc + weight, 0);

  const weightedSum = grades.reduce((acc, grade, index) => {
    return acc + grade * WEIGHTS[index];
  }, 0);

  return weightedSum / sumOfWeights;
}

function processExam(currentAverage: number, examGrade: number): void {
  console.log(`Nota do exame: ${examGrade.toFixed(1)}`);
  const finalAverage: number = (currentAverage + examGrade) / 2.0;
  if (finalAverage >= 5.0) {
    console.log("Aluno aprovado.");
  } else {
    console.log("Aluno reprovado.");
  }
  console.log(`Media final: ${finalAverage.toFixed(1)}`);
}

function truncateDecimal(value: number): number {
  return Math.floor(value * 10) / 10;
}

main();
